internal/k8s/reconciler/factory.go

```diff
@@ -0,0 +1,99 @@
+package reconciler
+
+import (
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/gatewayclient"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/state"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/service"
+	apigwv1alpha1 "github.com/hashicorp/consul-api-gateway/pkg/apis/v1alpha1"
+	"github.com/hashicorp/go-hclog"
+	gw "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+type Factory struct {
+	controllerName string
+	logger         hclog.Logger
+	client         gatewayclient.Client
+	deployer       *GatewayDeployer
+
+	resolver service.BackendResolver
+}
+
+type FactoryConfig struct {
+	ControllerName string
+	Logger         hclog.Logger
+	Client         gatewayclient.Client
+	Deployer       *GatewayDeployer
+
+	// get rid of this when validators are added
+	Resolver service.BackendResolver
+}
+
+func NewFactory(config FactoryConfig) *Factory {
+	return &Factory{
+		controllerName: config.ControllerName,
+		logger:         config.Logger,
+		client:         config.Client,
+		deployer:       config.Deployer,
+		resolver:       config.Resolver,
+	}
+}
+
+type NewGatewayConfig struct {
+	Gateway         *gw.Gateway
+	Config          apigwv1alpha1.GatewayClassConfig
+	State           *state.GatewayState
+	ConsulNamespace string
+}
+
+func (f *Factory) NewGateway(config NewGatewayConfig) *K8sGateway {
+	gatewayState := config.State
+	if gatewayState == nil {
+		gatewayState = state.InitialGatewayState(config.Gateway)
+	}
+	return f.initializeGateway(&K8sGateway{
+		Gateway:         config.Gateway,
+		GatewayState:    gatewayState,
```

Comment:
Any reason `GatewayState` should be exported on K8sGateway? That seems like something that should be private

Reply:
So, the state is actually going to require serialization alongside the actual gateway definitions once we can do the full serializing. Basically a "stored" gateway is the definition + its bound state, so just to keep with common Go conventions, made anything that's "serializable" exported.